### DISCUSSION TITLE
[Feature]: Select images in Immich using native system tools

### DISCUSSION THREAD

**Discussion by gh3597** [immich-app/immich#1641] [Feature Request]
### Feature detail

Ability to select photos from Immich to use on the device, e.g. composing a tweet in Android I can go to "More..." and select Dropbox, Nextcloud, Seafile etc. 

Alternatively have the Immich images show up as a system folder so I could select them using Files/3rd party file explorer?

[<img src="https://user-images.githubusercontent.com/118667161/202902580-11544639-759d-49dd-aef0-43e29bd2ab99.jpg" width=25% height=25%>](https://user-images.githubusercontent.com/118667161/202902580-11544639-759d-49dd-aef0-43e29bd2ab99.jpg)

### Platform

Mobile App

**Comment by bo0tzz** [comment]
I closed #1045 as a duplicate of this. It has some useful additional detail.

**Comment by Olliek95** [comment]
Any update on this? Would be great to have this feature. :)

**Comment by ibrahim-mubarak** [comment]
I think Documents Provider is the right thing to use here
https://developer.android.com/reference/android/provider/DocumentsProvider

**Comment by rosaLux161** [comment]
Any updates on this? That's an important feature IMHO. :)

**Comment by altispeedkschmidt** [comment]
What would it take to get something like this implemented? Would it be possible to put like a bounty on this of like hey if someone added this id pay $X?

**Reply by bo0tzz** [reply]
We prefer not to do bounties as they tend to misalign incentives.

**Reply by altispeedkschmidt** [reply]
Sounds good if I where able to higher a dev that would push the development upstream would you guys be willing to merge the changes?

**Reply by alextran1502** [reply]
For this feature, yes

**Comment by Jackinatox** [comment]
Why is this feature closed? When I want to choose an image, I only have google photos and files as options. When i choose Files under "other locations" i also cant find immich. If this isnt implemented i will give it a try, but i dont have high hopes

**Reply by bo0tzz** [reply]
It's not closed, what gave you that impression?

**Reply by Jackinatox** [reply]
Oh, mhm I came from the [Feature](https://github.com/immich-app/immich/issues/992), you closed and converted to a discussion so i thought it was implemented. 
Ok, then i will give it a shot, but i never did any android/flutter development...

**Comment by Olliek95** [comment]
I'd also still really like to see this implemented. I really like all the features Immich offers but it surprises me that something like this is not implemented yet. I often have to resort to using Google Photos only to select photos instead of Immich, which is a bummer.

**Comment by hasmar04** [comment]
While Immich cannot yet be assigned as a Cloud Media Provider due to Google's lengthy pilot program, once that program ends it is very related to this feature - https://developer.android.com/guide/topics/providers/cloud-media-provider
It allows Immich photos to appear in the native photo picker

**Reply by DocX** [reply]
Thanks for the update

**Reply by stavros-k** [reply]
Does iOS have any similar thing? I couldn't find anything online. But given it's apple they will add this 3 years after Android have passed the pilot program.

**Reply by griffi-gh** [reply]
Cloud Media Provider is a separate feature from Document provider.  
Ideally we should implement both, but just Document provider is enough for now.

**Reply by enzovic** [reply]
@griffi-gh You are absolutely right that DocumentsProvider is the correct immediate target. While CloudMediaProvider is the future for the Photo Picker, a robust DocumentsProvider solves the wider integration gap (files, emails, generic editors) immediately.
I have conducted a deep dive into the architectural requirements for bridging the Flutter/Native gap for this feature. Below is a technical proposal to implement this without degrading the user experience (ANRs) or compromising security.
1. The Core Challenge: "Activity-less" Execution
The DocumentsProvider is instantiated by the Android System (via ContentResolver), often while the main Immich Flutter activity is terminated.
 * Constraint: We cannot rely on the Dart VM, MethodChannels, or the existing Dart networking layer. Spinning up the Flutter engine for every file query would introduce unacceptable latency and UI stutter in the system picker.
 * Solution: The provider must be a pure Kotlin native component with an independent networking stack (Retrofit/OkHttp).
2. Proposed Architecture
A. Authentication: "Explicit Credential Mirroring"
We cannot safely read the flutter_secure_storage data from the native side because the plugin now utilizes internal implementations (Tink/AES) that are subject to change, risking a break on plugin updates.
 * Strategy: Implement a NativeAuthHelper. Upon a successful login or token refresh in Flutter, we explicitly write the apiUrl and accessToken to a native EncryptedSharedPreferences instance (Jetpack Security).
 * Benefit: Decouples the native provider from Flutter ecosystem changes while maintaining hardware-backed security.
B. Virtual File System (VFS) & IDs
To prevent TransactionTooLargeException and UI lags, we cannot present a flat list of 50k+ assets. We must project a virtual hierarchy.
 * Structure: Root -> Timeline -> Year -> Month -> Assets.
 * ID Schema: Custom synthetic IDs (e.g., MONTH:2024-02) will map the VFS to API queries. When queryChildDocuments receives a "Month" ID, it hits the Immich API for assets in that date range.
C. Data Transport & Streaming
The openDocument method requires different strategies for images vs. video to ensure compatibility with external apps.
 * Images: Use ParcelFileDescriptor.createReliablePipe(). A background thread streams bytes from the OkHttp response directly to the pipe.
 * Videos (The Seekability Trap): Pipes are not seekable. External video editors/players often fail if they cannot seek.
   * Strategy: "Download-to-Cache". For video MIME types, we download the file to context.cacheDir first, then return a read-only PFD to the actual file. This ensures full seekability.
D. Networking & Threading
 * Stack: Retrofit + OkHttp (with an Interceptor for the Bearer token).
 * Safety: Strict usage of Dispatchers.IO. We must not assume queryChildDocuments is called on a worker thread.
 * Offline Handling: "Fail-Fast" check on ConnectivityManager to prevent the System Picker from freezing/hanging on timeouts.
3. Implementation Plan
I propose breaking this into 3 PRs to manage complexity:
 * Foundation: Implement NativeAuthHelper (Flutter side) and EncryptedSharedPreferences (Native side) to ensure credentials are available to Kotlin.
 * The Browser: Implement the DocumentsProvider skeleton, VFS logic, and Retrofit client (Metadata only).
 * The Stream: Implement the Pipe (Image) and Cache (Video) streaming logic for openDocument.
This architecture also lays the groundwork for a future CloudMediaProvider, as the native networking and auth synchronization logic would be shared.

**Comment by kaysond** [comment]
See also #16445 - iOS has a similar integration through iMessages

**Reply by enzovic** [reply]
I don't think that many people use iPhones... 
I'm still waiting for the feature to be implemented in Android/Samsung. It's a pain to have to manually download a picture/video, every time I want to share it with other apps.

**Reply by kaysond** [reply]
[Citation Needed]
